sdks/java/core/src/main/java/org/apache/beam/sdk/coders/RowCoderGenerator.java

```diff
@@ -425,7 +538,7 @@ static Row decodeDelegate(
         // in which case we drop the extra fields.
         if (encodingPos < coders.length) {
           int rowIndex = encodingPosToIndex[encodingPos];
-          if (nullFields.get(rowIndex)) {
+          if (nullFields.get(encodingPos)) {
```

Comment:
was this a bug? rowIndex and encodingPos looks different

Reply:
Yes this along with other nullfields fix above is the purpose of this PR to fix #32388 .
The stack trace and synchronization changes were added as the initial belief was that encoded corruption was due to late overrides arriving.  Since that could still be an issue, I think we should keep those changes but I can separate them to a separate PR if you'd prefer.

Reply:
Note that unless there are encoding overrides rowIndex and encodingPos are equal. But the improved unit tests catch the issue, previous tests with encoding overrides didn't have null fields and thus missed it.